sphinx/config.py

```diff
@@ -465,6 +465,19 @@ def __getstate__(self) -> dict:
         for name, opt in self._options.items():
             real_value = getattr(self, name)
             if not is_serializable(real_value):
+                if opt.rebuild:
+                    # if the value is not cached, then any build that utilises this cache
+                    # will always mark the config value as changed,
+                    # and thus always invalidate the cache and perform a rebuild.
+                    logger.warning(
+                        __(
+                            'The value of config %r is not picklable and so cannot be cached.'
+                        ),
+                        name,
+                        type='config',
+                        subtype='cache',
```

Comment:
Do you think it's preferrable to always specify the subtype? One issue with *subtype* is that you have many subtypes to document and once documented, it's hard to change them.

Reply:
So I think at least in this instance, it is prefereable to have a more specific warning type, since this is probably a warning that advanced users may want to add to `suppress_warnings`,
i.e. they acknowledge that they are essentially breaking sphinx's caching mecahnisms, but still want to continue using the uncacheable values

More generally, you may interested to know that in myst-parser, I have a layer on top of sphinx warnings, to use an enum for storing all warning types used in the extension: https://github.com/executablebooks/MyST-Parser/blob/dfaf6d7abfdd988c74a1d6e1bbdc637e9f5b5bcb/myst_parser/warnings_.py#L11

Then I have a custom directive to auto-document these in: https://myst-parser.readthedocs.io/en/latest/configuration.html#build-warnings 😄 

(similar to my layer on top of `Config` that I just mentioned in https://github.com/sphinx-doc/sphinx/pull/12198#issuecomment-2020048581)

Reply:
TBH, especially with my addition of https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-show_warning_types, I feel listing the warning types in https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-suppress_warnings is not so useful,
as I would envisage the actual user workflow would be that they run their build with `show_warning_types = True`, then use the warning types emitted to decide what warning types they want to suppress.

Having a non-autogenerated list is obviously very prone to "desyncing" from the actual codebase

this is though not within the scope of this PR

Reply:
Mmh. Even with `show_warning_types`, I'm in favor of having a documented entry (possibly autogenerated). But I'm fine with using a subtype (although I don't like 'cache', I don't have a better alternative).